File: 1001-2000/1238-circular-permutation-in-binary-representation/1238-circular-permutation-in-binary-representation.java
class Solution {
    public List<Integer> circularPermutation(int n, int start) {
        int x=(2<<(n-1))-1;
        int c=0;
        List<Integer> a=new ArrayList<>();
        //a.add(x);
        int k=0,h=0;
        while(k<=x){
            h=k^(k>>1);
            if(h==start){
                break;
            }
            k++;
        }
        for(int i=k;i<=x;i++){
            c=i^(i>>1);
            a.add(c);
        }
        for(int i=0;i<k;i++){
            c=i^(i>>1);
            a.add(c);
        }
        return a;
    }
}

File: 1001-2000/1238-circular-permutation-in-binary-representation/1238-circular-permutation-in-binary-representation.cpp
class Solution {
public:
    vector<int> circularPermutation(int n, int start) {
        int size= 1<<n; 
        vector<int>ans(size); 
        int index=-1; 
        for(int i=0; i< size; i++) {
            ans[i]= (i ^i>>1);
            if(ans[i]==start) index= i;
        }
        reverse(ans.begin(), ans.begin()+index); 
        reverse(ans.begin()+index, ans.end());
        reverse(ans.begin(), ans.end()); 
        return ans; 
    }
};